test/framework/flux.go

```diff
@@ -64,13 +64,77 @@ func WithFlux(opts ...api.GitOpsConfigOpt) ClusterE2ETestOpt {
 	}
 }
 
+func WithClusterUpgradeGit(fillers ...api.ClusterFiller) ClusterE2ETestOpt {
+	return func(e *ClusterE2ETest) {
+		e.ClusterConfigB = e.customizeClusterConfig(e.clusterConfigGitPath(), fillers...)
+
+		// TODO: the gitopsconfig object is generated from api.NewGitOpsConfig each time
+		// instead of marshalling from the file in git repo. Flux requires namespace always
+		// exist for kustomization. Need to refactor this to read gitopsconfig from file.
+		if e.GitOpsConfig.GetNamespace() == "" {
+			e.GitOpsConfig.SetNamespace("default")
+		}
+	}
+}
+
 func withFluxRepositorySuffix(suffix string) api.GitOpsConfigOpt {
 	return func(c *v1alpha1.GitOpsConfig) {
 		repository := c.Spec.Flux.Github.Repository
 		c.Spec.Flux.Github.Repository = fmt.Sprintf("%s-%s", repository, suffix)
 	}
 }
 
+func (e *ClusterE2ETest) UpgradeThroughGitOps(clusterOpts []ClusterE2ETestOpt) {
+	e.upgradeThroughGitOps(clusterOpts)
+}
+
+func (e *ClusterE2ETest) upgradeThroughGitOps(clusterOpts []ClusterE2ETestOpt) {
+	ctx := context.Background()
+	e.initGit(ctx)
+
+	if err := e.validateInitialFluxState(ctx); err != nil {
+		e.T.Errorf("Error validating initial state of cluster gitops system: %v", err)
+	}
+
+	for _, opt := range clusterOpts {
+		opt(e)
+	}
+
+	e.buildClusterConfigFileForGit()
+
+	if err := e.pushConfigChanges(); err != nil {
+		e.T.Errorf("Error pushing local changes to remote git repo: %v", err)
+	}
+	e.T.Logf("Successfully updated version controlled cluster configuration")
+
+	if err := e.validateWorkerNodeUpdates(ctx); err != nil {
```

Comment:
did you have plans for these specific validation methods in future work? I feel like these should be moved out of `upgradeThroughGitOps` and into their own separate validation methods, separate from the initialization of the repository/cluster config generation/git push

thinking that as a test user I may want to modify the cluster in some way after performing the upgrade via git, before performing the validation (for example, modifying the cluster directly and then validating that it returns to the expected state, just as one idea; or running a CLI upgrade flow and then validating it)

for now, even just consider bumping the validations up to the public `UpgradeThroughGitops` method so we set up the distinction between the initialization/push and the validation

Reply:
great point and I agree we should. This validation method should be shared among e2e cli create/upgrade and gitops.

added in https://github.com/aws/eks-anywhere/issues/396